// ts-json-schema-generator/dist/src/TypeFormatter/IntersectionTypeFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IntersectionTypeFormatter = void 0;
const ArrayType_js_1 = require("../Type/ArrayType.js");
const IntersectionType_js_1 = require("../Type/IntersectionType.js");
const TupleType_js_1 = require("../Type/TupleType.js");
const allOfDefinition_js_1 = require("../Utils/allOfDefinition.js");
const uniqueArray_js_1 = require("../Utils/uniqueArray.js");
class IntersectionTypeFormatter {
    childTypeFormatter;
    constructor(childTypeFormatter) {
        this.childTypeFormatter = childTypeFormatter;
    }
    supportsType(type) {
        return type instanceof IntersectionType_js_1.IntersectionType;
    }
    getDefinition(type) {
        const dependencies = [];
        const nonArrayLikeTypes = [];
        for (const t of type.getTypes()) {
            if (t instanceof ArrayType_js_1.ArrayType || t instanceof TupleType_js_1.TupleType) {
                dependencies.push(this.childTypeFormatter.getDefinition(t));
            }
            else {
                nonArrayLikeTypes.push(t);
            }
        }
        if (nonArrayLikeTypes.length) {
            dependencies.push(nonArrayLikeTypes.reduce((0, allOfDefinition_js_1.getAllOfDefinitionReducer)(this.childTypeFormatter), {
                type: "object",
                additionalProperties: false,
            }));
        }
        return dependencies.length === 1 ? dependencies[0] : { allOf: dependencies };
    }
    getChildren(type) {
        return (0, uniqueArray_js_1.uniqueArray)(type
            .getTypes()
            .reduce((result, item) => [...result, ...this.childTypeFormatter.getChildren(item)], []));
    }
}
exports.IntersectionTypeFormatter = IntersectionTypeFormatter;
