// ts-json-schema-generator/dist/src/TypeFormatter/TupleTypeFormatter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TupleTypeFormatter = void 0;
const ArrayType_js_1 = require("../Type/ArrayType.js");
const OptionalType_js_1 = require("../Type/OptionalType.js");
const RestType_js_1 = require("../Type/RestType.js");
const TupleType_js_1 = require("../Type/TupleType.js");
const notNever_js_1 = require("../Utils/notNever.js");
const uniqueArray_js_1 = require("../Utils/uniqueArray.js");
function uniformRestType(type, check_type) {
    const inner = type.getType();
    return ((inner instanceof ArrayType_js_1.ArrayType && inner.getItem().getId() === check_type.getId()) ||
        (inner instanceof TupleType_js_1.TupleType &&
            inner.getTypes().every((tuple_type) => {
                if (tuple_type instanceof RestType_js_1.RestType) {
                    return uniformRestType(tuple_type, check_type);
                }
                else {
                    return tuple_type?.getId() === check_type.getId();
                }
            })));
}
class TupleTypeFormatter {
    childTypeFormatter;
    constructor(childTypeFormatter) {
        this.childTypeFormatter = childTypeFormatter;
    }
    supportsType(type) {
        return type instanceof TupleType_js_1.TupleType;
    }
    getDefinition(type) {
        const subTypes = type.getTypes().filter(notNever_js_1.notNever);
        const requiredElements = subTypes.filter((t) => !(t instanceof OptionalType_js_1.OptionalType) && !(t instanceof RestType_js_1.RestType));
        const optionalElements = subTypes.filter((t) => t instanceof OptionalType_js_1.OptionalType);
        const restType = subTypes.find((t) => t instanceof RestType_js_1.RestType);
        const firstItemType = requiredElements.length > 0 ? requiredElements[0] : optionalElements[0]?.getType();
        const isUniformArray = firstItemType &&
            requiredElements.every((item) => item.getId() === firstItemType.getId()) &&
            optionalElements.every((item) => item.getType().getId() === firstItemType.getId()) &&
            (!restType || uniformRestType(restType, firstItemType));
        if (isUniformArray) {
            return {
                type: "array",
                items: this.childTypeFormatter.getDefinition(firstItemType),
                minItems: requiredElements.length,
                ...(restType ? {} : { maxItems: requiredElements.length + optionalElements.length }),
            };
        }
        const requiredDefinitions = requiredElements.map((item) => this.childTypeFormatter.getDefinition(item));
        const optionalDefinitions = optionalElements.map((item) => this.childTypeFormatter.getDefinition(item));
        const itemsTotal = requiredDefinitions.length + optionalDefinitions.length;
        const additionalItems = restType ? this.childTypeFormatter.getDefinition(restType).items : undefined;
        return {
            type: "array",
            minItems: requiredDefinitions.length,
            ...(itemsTotal ? { items: requiredDefinitions.concat(optionalDefinitions) } : {}),
            ...(!itemsTotal && additionalItems ? { items: additionalItems } : {}),
            ...(!itemsTotal && !additionalItems ? { maxItems: 0 } : {}),
            ...(additionalItems && !Array.isArray(additionalItems) && itemsTotal
                ? { additionalItems: additionalItems }
                : {}),
            ...(!additionalItems && itemsTotal ? { maxItems: itemsTotal } : {}),
        };
    }
    getChildren(type) {
        return (0, uniqueArray_js_1.uniqueArray)(type
            .getTypes()
            .filter(notNever_js_1.notNever)
            .reduce((result, item) => [...result, ...this.childTypeFormatter.getChildren(item)], []));
    }
}
exports.TupleTypeFormatter = TupleTypeFormatter;
